packages/flutter/lib/src/scheduler/binding.dart

```diff
@@ -214,6 +215,8 @@ abstract class SchedulerBinding extends BindingBase with ServicesBinding {
         timeDilation = value;
       }
     );
+
+    LoggingService.instance.initServiceExtensions(registerServiceExtension);
```

Comment:
this seems dramatically overengineered for the use cases.

Alternative strawman proposals:

## Proposal 1

Just use `Timeline` in `dart:developer`. It already supports naming events and taking arbitrary arguments.

## Proposal 2

The foundation binding exposes a single global method that logs an event, say `debugLogEvent()`. This takes a string and a closure that matches this typedef:

```
  typedef DebugLogEventCallback = DebugLogEvent Function();
```

`DebugLogEvent` is a class:

```dart
abstract class DebugLogEvent {
  const DebugLogEvent();
  @protected
  Foo serialize();
}
```

...where `Foo` is whatever type you need to send data over the service protocol.

The foundation binding registers a service protocol handler for enabling or disabling logging of particular sets of data. This takes a string and a boolean. When the boolean is true, the string is added to a set. When it's false, the string is removed from the set.

When `debugLogEvent` is called with a string that is in the set, the closure is called, and the resulting object's `serialize` method is called and the results sent over the service protocol to any listening IDEs.

Someone who wants to send data subclasses `DebugLogEvent` and defines a constant for the name of the class of data. They then call `debugLogEvent` like this:

```dart
  debugLogEvent(logGestures, () => new DebugGestureLogEvent(foo.x, foo.y, bar.baz));
```

...where `foo` and `bar` are object whose data is getting logged.

Reply:
This is great.  Thanks @Hixie!

I'm leaning towards `#2` but wonder if really need an intermediate log object?  That is, why not just have the closure return `Foo`?

Regarding `Foo`, I'm guessing this will just be `Object` with the expectation that it be json-encodable, since this is the payload for [postEvent](https://api.dartlang.org/stable/2.0.0/dart-developer/postEvent.html).

In the end, I wonder if we couldn't replace:

    typedef DebugLogEventCallback = DebugLogEvent Function();

with

    typedef DebugLogEventCallback = Object Function();

making logging calls look like:

```
  debugLogEvent(logGestures, () => <String, int> {
    'x' : x,
    'y' : y,
    'z' : z,
  });
```

or as simple as:

```
  debugLogEvent('flutter.log.hello', () => 'hello world');
```

Reply:
Also, I wonder if we shouldn't take the approach that `package:logging` takes and support closures _or_ objects in the `debugLogEvent` call.

See for example https://github.com/dart-lang/sdk/issues/31113 on the expense of closure allocation.

Reply:
I guess you could also have the closure return Foo, that just seems less pretty.

Please don't take objects in `debugLogEvent`; the whole point is to avoid doing any work when the logging is turned off. I'd very much like to even avoid allocating the closures but I don't know how to do that in a reasonable way here, short of requiring the pattern be:

```dart
if (debugShouldLogEvent('foo'))
  debugLogEvent('foo', ...);
```

We already pay the cost of closure allocation for `assert` all over the place, so if it's a hot path then it's already a hot path and already would benefit from optimisation, so using it more isn't adding a new problem.

I would very much like us to keep these log lines to a strict minimum, because of the cost. As I noted earlier, we already have had to remove some asserts because they were too expensive.

Reply:
(actually, the code above would also not remove the closure, because it'd have to be:

```dart
  assert(() {
    if (debugShouldLogEvent('foo'))
      debugLogEvent('foo', ...);
    return true;
  }());
```

...so there's no win there.)

Reply:
I think longer term a kernel transform could strip out all the logging (and `profile()`) code pretty well. @jacob314 and @pq have discussed such.

Reply:
> I guess you could also have the closure return Foo, that just seems less pretty.

My worry is just that the first proposal would lead to a proliferation of `DebugLogEvent` classes that were just boilerplate.  For example, 

```dart
class _DebugGestureLogEvent extends DebugLogEvent {
  final int x;
  final int y;
  final int z;
  const _DebugGestureLogEvent(this.x, this.y, this.z);

  @override
  Object serialize() => <String, int> {
      'x' : x,
      'y' : y,
      'z' : z,
    };
}
```

feels like kind of a lot to allow us to write:

```dart
debugLogEvent(logGestures, () => new DebugGestureLogEvent(foo.x, foo.y, bar.baz));
```

Admittedly, that's nice nicer than:

```dart
  debugLogEvent(logGestures, () => <String, int> {
    'x' : x,
    'y' : y,
    'z' : z,
  });
```

but is it worth the class?  But maybe I'm missing something... and there are other benefits I'm overlooking?

> Please don't take objects in `debugLogEvent`.

👍

As for closure allocations you make a good point about paying the price in the wrapping `assert`.   

> I would very much like us to keep these log lines to a strict minimum, because of the cost. As I noted earlier, we already have had to remove some asserts because they were too expensive.

Understood.  We'll definitely want to be judicious and yeah, maybe on hot paths consider an `if (debugShouldLogEvent('foo'))` idiom.

Reply:
> I think longer term a kernel transform could strip out all the logging (and profile() / assert()) code pretty well.

👍 

Absolutely.  Short to medium term we should do this since it will be straight-forward (if I understand correctly).  That said, we still will need to be careful to not destroy debug perf...